jukepoot/puppet_parser.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/python
+"""A parser of /var/log/puppet.log.
+
+Imports argparse for parsing arguments and then the script
+accepts file path and year as optional argument. 
+"""
+
+
+import argparse 
+from datetime import datetime
+import dateutil.parser
+import re
+import collections
+import os
+
+
+def main():
+	parser = argparse.ArgumentParser()
+	parser.add_argument(help="parse the given filename", type=str, dest="filename", action="store")
+	parser.add_argument("-y","--year", help="parse file with given year", type=int, dest="year", action="store")
+	args = parser.parse_args()
+	if args.year:
+		print("year to use: " + str(args.year))
+		year = args.year
+
+	else:
+		year = datetime.utcfromtimestamp(os.stat(args.filename).st_mtime).year
+	
+	puppet_parser(args.filename,year)
+
+
+def puppet_parser(filepath, year):
+	"""Accepts the puppet file path and parses it out by calling other functions.
+
+	"""
+	#TODO use year from last modification
+
+	# pattern to match from log file and group each finding for referencing
+	# pattern looks for a time foormat (x:x:x)any text after which is
+	# followed by the process and so on 
+	pattern = re.compile(r'(^.+:..:..)\s(\S+)\s(\S+)(\[\d+\]):\s(.+$)')
+	#group 1 = date
+	#group 2 = hostname
+	#group 3 = proccess
+	#group 4 = PID
+	#group 5 = message
+	
+	# add to list for counter purposes 
+	message_list = []
+	proccess_list =[]
+	hostname_list= []
+	processID_list = []
+	
+	
+	f = open(filepath, "rU")
+	for line in f:
+
```

Comment:
OK... so I would do two things

a) split file parsing and log parsing in two parts... as in have this puppet parser open up the file, read a line in and then call a ParseLine function or something like that.... for each line, that way you'll split up the log parsing from the generic file handling

b) do

```
match = pattern.match(line)

if not match:
  continue (or return if this is in a function)

date_string = match.group(1)
date = dateutil.parser.parse(date_string)

if date.year == year:
  # do something

```

Also I'm not sure if I follow this logic... you pass in a year as an argument, and you just check to see if the year is the same as the one you pass in?

The whole problem with this is the fact that the puppet.log file doesn't contain an year... so there is no year there...

```
In [1]: import dateutil.parser

In [2]: d = dateutil.parser.parse('Jul 25 20:24:29')

In [3]: d.year
Out[3]: 2018
```

You see the dateutil parser will just guess the year, since the text doesn't contain it... you'll need to SET the year based on either the default one or the one that gets passed in as an argument

Reply:
no, so . the point is that you need to parse the date, and if the date is:
Jul 25 20:24:29

You can't really parse this date, since you don't have any information about what year it is... thus we do these two "patch works":
1. read the year through an argument
2. if no year is passed then use the last modification time of the file.

so you do this... you get the year through argparse or through file modification... but then you need to use that year to build the date... what you are doing here is

+ build a date object without providing the year -> dateutil will guess and just use the current year
+ then check if the year that dateutil has no idea what is and is just guessing to compare against the year that you calculated

So basicalyl what you are doing is that irrelevant of the file you are parsing, all entries will have CURRENT year that the script is running, but you only include entries from the provided year or file modification... so let's say that the file was last modified in december, and you run the script in january.... then you get no entries, whereas you shuold get them all

(and BTW file modification is not a good time to use, since if the file gets created in December, and then you parse it in january, the last modification time is most likely in january...then you interpret all timestamps as the current year, which is swrong since the first entries are all from last year... do you understand that? ... so you'll need to increment the year if it pushes over year boundary)... so if the entry starts in december and then there are records from january, then the entries from jan are year+1 from the ones in december).....

so the idea here is that file CREATION or year passed in as an argument is the starting year of the file, so all date entries that are parsed use that as the year, since there is no year in the dates, and then you need to think about year boundaries to increment....